### PR TITLE
Bitbucket integration

### PR DESCRIPTION
**Pull request by nid90**
**Closes:** #639

### Differences from other VCS Integration (GitHub mainly)

- Cherry picking cannot be implemented properly, so the continuous back-merge feature is disabled
- The artifact download depends on the Bitbucket pipeline uploading the file to the Downloads section of the repository with a particular filename pattern `build-<PIPELINE_UUID>`
- Author email is extracted from raw author data in a best-guess manner and other author details like URL, login are not available

#### Creation

- schema
- user flow to add integration
- metadata about the integration
- configuration to choose repository

#### Add necessary methods to talk to the integration
- covers branches, tags , commits, pull requests, releases, diffs etc.
- full list here: https://github.com/tramlinehq/tramline/issues/639#issue-2502616029

#### Events from webhook
- handle push events
- handle pull request events 

#### Creation

- schema
- user flow to add integration
- metadata about the integration

#### Add necessary methods to talk to the integration
- covers workflows, artifacts etc.
- full list here: https://github.com/tramlinehq/tramline/issues/639#issue-2502616029
- fetching artifact from the integration can change somewhat significantly between integrations, and so the logic is encapsulated in `WorkflowProcessors::<Integration>::WorkflowRun`and cannot be replaced with a unifying transformation

### Remaining TODOs:

- [x] Configure the workspace correctly
- [x] Figure out how to fetch a list of pipeline steps to allow the user to configure internal and RC workflow
- [x] Artifact download needs to be handled in the pipeline itself
- [x] BItbucket commits in all forms (webhook event and API) have parents information, no need to fetch it again during push processing
~- [ ] Cherry-pick PR~
- [x] Handle author email for commits

### CODE REVIEW

**Comment by prabhanshuguptagit** [comment]
## VCS APIs 
```ruby
# Use an existing integration
integration = BitbucketIntegration.first
org = "tramline"
repo_slug = "ueno"

integration.send(:reset_tokens!)
bb = Installations::Bitbucket::Api.new(integration.oauth_access_token, org)

# webhooks
hook_id = bb.create_repo_webhook!(repo_slug, "#{ENV["WEBHOOK_HOST_NAME"]}/bitbucket/events/1234", BitbucketIntegration::WEBHOOK_TRANSFORMATIONS)["id"]
bb.find_webhook(repo_slug, hook_id, BitbucketIntegration::WEBHOOK_TRANSFORMATIONS)
bb.update_repo_webhook!(repo_slug, hook_id, "#{ENV["WEBHOOK_HOST_NAME"]}/bitbucket/events/12345", BitbucketIntegration::WEBHOOK_TRANSFORMATIONS)

# branches
bb.create_branch!(repo_slug, "main", "test-branching-8") # from a branch
bb.create_branch!(repo_slug, "test-branch-creation", "test-branching-9", source_type: :tag) # from a tag
bb.create_branch!(repo_slug, "2852d77475aa1f95a60e67f4c28081aa1835068d", "test-branching-10", source_type: :commit) # from a commit
bb.branch_exists?(repo_slug, "test-branching-10")

# tags
bb.create_tag!(repo_slug, "test-tagging-2", "2852d77475aa1f95a60e67f4c28081aa1835068d")
bb.tag_exists?(repo_slug, "test-tagging-2")

# PRs
pr = bb.create_pr!(repo_slug, "main", "test-webhooks", "Test PR from API #2", "A description of my PR", BitbucketIntegration::PR_TRANSFORMATIONS)
bb.find_pr(repo_slug, "main", "test-webhooks", BitbucketIntegration::PR_TRANSFORMATIONS)
bb.get_pr(repo_slug, pr["number"], BitbucketIntegration::PR_TRANSFORMATIONS)
bb.get_pr(repo_slug, pr["number"].to_s, BitbucketIntegration::PR_TRANSFORMATIONS)
bb.merge_pr!(repo_slug, pr["number"])

# commits
bb.get_commit("ueno", "1db233942778c67f398ddc1c2ae0f4df27f5ebd8", BitbucketIntegration::COMMITS_TRANSFORMATIONS)
bb.head("ueno", "test-1webhooks", commit_transforms: BitbucketIntegration::COMMITS_TRANSFORMATIONS)
bb.commits_between("ueno", "6ce9089", "1db233942778c67f398ddc1c2ae0f4df27f5ebd8", BitbucketIntegration::COMMITS_TRANSFORMATIONS)
```

File: app/models/bitbucket_integration.rb [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  }
+
+  COMMITS_TRANSFORMATIONS = {
+    url: [:links, :html, :href],
+    commit_hash: :hash,
+    message: :message,
+    # TODO: author information is not available when user does
+    # not match on atlassian
+    author_name: [:author, :user, :display_name],
+    # TODO: email is not available
```
Comment (prabhanshuguptagit) [inline app/models/bitbucket_integration.rb:118]: Bitbucket doesn't provide the author email, should we instead make use of `[:author :raw] => Brodie Rao <a@b.c>`?

File: app/models/bitbucket_integration.rb [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+  def branch_url(branch_name)
+    "https://bitbucket.org/#{code_repository_name}/branch/#{branch_name}"
+  end
+
+  def tag_url(tag_name)
+    "https://bitbucket.org/#{code_repository_name}/src/#{tag_name}"
+  end
+
+  def compare_url(to_branch, from_branch)
```
Comment (kitallis) [inline app/models/bitbucket_integration.rb:203]: Something like this

https://bitbucket.org/tramline/ueno/branch/r/bb-fuill-train/2024-09-16?dest=r%2Fbb-fuill-train%2F2024-09-16

File: app/models/bitbucket_integration.rb [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+  end
+
+  def get_workflow_run(pipeline_id)
+    with_api_retries { installation.get_pipeline(code_repository_name, pipeline_id) }
+  end
+
+  def get_artifact_v2(_, _, external_workflow_run_id:)
+    raise Integration::NoBuildArtifactAvailable if external_workflow_run_id.blank?
+
+    artifact_name = "build-#{external_workflow_run_id}".gsub(/{/, "").gsub(/}/, "")
```
Comment (kitallis) [inline app/models/bitbucket_integration.rb:311]: Write a comment explaining these braces oddity.

File: app/models/bitbucket_integration.rb [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+  def trigger_workflow_run!(ci_cd_channel, _branch_name, inputs, commit_hash = nil)
+    with_api_retries do
+      res = installation.trigger_pipeline!(code_repository_name, ci_cd_channel, inputs, commit_hash, WORKFLOW_RUN_TRANSFORMATIONS)
+      res.merge(ci_link: "https://bitbucket.org/#{code_repository_name}/pipelines/results/#{res[:number]}")
+    end
+  end
+
+  def retry_workflow_run!(_ci_ref)
+    raise Integrations::UnsupportedAction
```
Comment (kitallis) [inline app/models/bitbucket_integration.rb:293]: Implement this

File: app/models/bitbucket_integration.rb [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+  end
+
+  def events_url(params)
+    if Rails.env.development?
+      bitbucket_events_url(host: ENV["WEBHOOK_HOST_NAME"], **params)
+    else
+      bitbucket_events_url(host: ENV["HOST_NAME"], protocol: "https", **params)
+    end
+  end
+end
```
Comment (kitallis) [inline app/models/bitbucket_integration.rb:379]: Use `Linkable` concern to simplify.

**Review by kitallis — CHANGES_REQUESTED** [review]
Don't forget to add Prabhanshu to contributors in the merge commit.

File: app/models/bitbucket_integration.rb (hunk 4 above)
Comment (nid90) [reply app/models/bitbucket_integration.rb:293]: Bitbucket doesn't support retrying the existing pipeline, instead, It creates a new pipeline. So, this method cannot be implemented.

We handle this by letting `WorkflowRun` know by making `BitbucketIntegration` as not `workflow_retriable`.

File: app/libs/installations/bitbucket/api.rb [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+              key: "VERSION_CODE",
+              value: inputs[:version_code]
+            }
+          ]
+        }
+      }
+
+      execute(:post, PIPELINES_URL.expand(workspace:, repo_slug:).to_s, params)
+        .then { |response| Installations::Response::Keys.transform([response], transforms) }
+        .first
```
Comment (nid90) [inline app/libs/installations/bitbucket/api.rb:306]:
```suggestion
        .first
        .then { |run| run&.presence || raise(Installations::Error.new("Could not find the workflow run", reason: :workflow_run_not_found)) }
```

**Pull request merged**